sqlmesh/core/snapshot/evaluator.py

```diff
@@ -989,40 +999,60 @@ def _promote_snapshot(
                 table_mapping=table_mapping,
                 runtime_stage=RuntimeStage.PROMOTING,
             )
-            _evaluation_strategy(snapshot, adapter).promote(
-                table_name=table_name,
-                view_name=view_name,
-                model=snapshot.model,
-                environment=environment_naming_info.name,
-                snapshots=snapshots,
-                **render_kwargs,
-            )
 
-            snapshot_by_name = {s.name: s for s in (snapshots or {}).values()}
-            render_kwargs["snapshots"] = snapshot_by_name
-            adapter.execute(snapshot.model.render_on_virtual_update(**render_kwargs))
+            with (
+                adapter.transaction(),
+                adapter.session(snapshot.model.render_session_properties(**render_kwargs)),
+            ):
+                _evaluation_strategy(snapshot, adapter).promote(
+                    table_name=table_name,
+                    view_name=view_name,
+                    model=snapshot.model,
+                    environment=environment_naming_info.name,
+                    snapshots=snapshots,
+                    **render_kwargs,
+                )
+
+                snapshot_by_name = {s.name: s for s in (snapshots or {}).values()}
+                render_kwargs["snapshots"] = snapshot_by_name
+                adapter.execute(snapshot.model.render_on_virtual_update(**render_kwargs))
 
         if on_complete is not None:
             on_complete(snapshot)
 
     def _demote_snapshot(
         self,
-        snapshot: SnapshotInfoLike,
+        snapshot: Snapshot,
         environment_naming_info: EnvironmentNamingInfo,
+        snapshots: t.Optional[t.Dict[SnapshotId, Snapshot]],
+        table_mapping: t.Optional[t.Dict[str, str]],
+        deployability_index: t.Optional[DeployabilityIndex],
         on_complete: t.Optional[t.Callable[[SnapshotInfoLike], None]],
     ) -> None:
-        adapter = (
-            self.get_adapter(snapshot.model_gateway)
-            if environment_naming_info.gateway_managed
-            else self.adapter
-        )
-        view_name = snapshot.qualified_view_name.for_environment(
-            environment_naming_info, dialect=adapter.dialect
-        )
-        _evaluation_strategy(snapshot, adapter).demote(view_name)
+        if snapshot.is_model:
```

Comment:
Can we just do `if not snapshot.is_model: return`? No point in nesting the entire body of the function.